Python/compile.c

```diff
@@ -3707,7 +4033,13 @@ compiler_nameop(struct compiler *c, location loc,
 
     op = 0;
     optype = OP_NAME;
-    scope = _PyST_GetScope(c->u->u_ste, mangled);
+    if (c->u->u_scope_type == COMPILER_SCOPE_CLASS &&
+        _PyUnicode_EqualToASCIIString(name, "__classdict__")) {
+        scope = CELL;
+    }
```

Comment:
It's needed here but not for `__class__` because unlike `__class__`, `__classdict__` needs to work while the class body is executing. Without the check I get:

```
>>> class X:
...     def f[T](self, x: int): pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in X
  File "<stdin>", line 2, in <generic parameters of f>
NameError: cannot access free variable '__classdict__' where it is not associated with a value in enclosing scope
```

This also happens (correctly, in this case) if you call `super()` in a function class body:

```
>>> class X:
...     def meth(self): super().meth()
...     meth(None)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in X
  File "<stdin>", line 2, in meth
NameError: cannot access free variable '__class__' where it is not associated with a value in enclosing scope. Did you mean: 'self.__class__'?
```

We need some special casing around this name because normally for class-scoped names we'll generate STORE_NAME, but for `__classdict__` we need to do STORE_DEREF.

As I wrote this I realized there's a simpler solution: we don't need to call compiler_nameop() at all when storing `__classdict__`, we can just generate a STORE_DEREF directly. I implemented that and it works.

Reply:
Ah, I didn't realize that this special case was for the storing to the cell in the class namespace, that makes sense! Just calling `compiler_nameop` directly for that scenario sounds good.